### PR TITLE
Extract router

### PR DESCRIPTION
**Pull request by nmathewson**

### CODE REVIEW

File: src/feature/relay/clientkey.c [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+ * Copyright (c) 2004-2006, Roger Dingledine, Nick Mathewson.
+ * Copyright (c) 2007-2020, The Tor Project, Inc. */
+/* See LICENSE for licensing information */
+
+/**
+ * @file clientkey.c
+ * @brief "Client"-related key management operations.
+ *
+ * These functions are used by clients and relay code paths alike, in spite of
+ * residing in feature/relay.  They should be refatored, moved, made
```
Comment (tlyu) [inline src/feature/relay/clientkey.c]: s/refatored/refactored/

File: src/feature/relay/descmgr.c [hunk 2]
```diff
@@ -0,0 +1,1744 @@
...
+  ri->cache_info.published_on = time(NULL);
+  /* get_onion_key() must invoke from main thread */
+  router_set_rsa_onion_pkey(get_onion_key(), &ri->onion_pkey,
+                            &ri->onion_pkey_len);
+
+  ri->onion_curve25519_pkey =
+    tor_memdup(&get_current_curve25519_keypair()->pubkey,
+               sizeof(curve25519_public_key_t));
+
+  /* For now, at most one IPv6 or-address is being advertised. */
```
Comment (teor2345) [inline src/feature/relay/descmgr.c]: Trac #32588 changed this code, and we just merged it.
Please rebase.

File: src/feature/relay/router.h [hunk 3]
```diff
@@ -73,74 +65,129 @@ uint16_t router_get_advertised_dir_port(const or_options_t *options,
...
+
+#define router_compare_to_my_exit_policy(addr, port) \
+  IGNORE2_RET((addr), (port), -1)
+#define router_my_exit_policy_is_reject_star() \
+  (1)
+
+#define should_refuse_unknown_exits(opts) \
+  IGNORE1_RET((opts), 1)
+
+/* TODO: These should be disabled by v3auth mode, not relay mode. */
```
Comment (teor2345) [inline src/feature/relay/router.h]: Is there another ticket for the dirauth change?

File: src/feature/relay/router.h [hunk 4]
```diff
@@ -73,74 +65,129 @@ uint16_t router_get_advertised_dir_port(const or_options_t *options,
...
+  IGNORE3_RET((opt), (addr), (cache), -1)
+
+#define router_new_address_suggestion(s, d) \
+  IGNORE2((s),(d))
+
+#define init_keys() \
+  init_keys_client()
+
+#define init_keys_common() \
+  (0)
```
Comment (teor2345) [inline src/feature/relay/router.h]: init_keys_common() creates the key mutex.

If we don't need the key mutex on clients, let's open another ticket to remove it?
(And if we do need it, let's leave it alone.)

Comment (nmathewson) [reply src/feature/relay/router.h]: I think that key_lock *is* currently removed on clients with this patch.  It is declared in router.c, and is only used there: it protects onion keys only.

Comment (teor2345) [reply src/feature/relay/router.h]: Can we have some comments around these changes, for the sake of future developers? :-'

**Pull request closed without merging**